Feature/co 1116 post/cards endpoint

ianbrown9475 opened the pull request.

ianbrown9475 committed:
Add POST functionality

ianbrown9475 committed:
Add POST functionality

ianbrown9475 committed:
Add POST functionality

ianbrown9475 committed:
Merge branch 'feature/CO-1116-POST/cards_endpoint' of https://github.com/osu-mist/slay-the-spire-cards-api into feature/CO-1116-POST/cards_endpoint

ianbrown9475 committed:
Merge branch 'feature/CO-1116-POST/cards_endpoint' of https://github.com/osu-mist/slay-the-spire-cards-api into feature/CO-1116-POST/cards_endpoint

ianbrown9475 committed:
Add better error checking

ianbrown9475 committed:
Factor cardValidator method

ianbrown9475 committed:
Remove unused dependencies

ianbrown9475 committed:
Remove unused annotations

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/cardsapi/resources/CardsResource.groovy:
```diff
@@ -65,4 +82,129 @@ class CardsResource extends Resource {
+                    "\'. Description must match pattern: " +
+                    regEx).build()
+        }
+        if(!(card.energy >= 0 && card.energy <= 999)) {
```
perhaps these numbers should be set as variables for the class? Then you could reference them in your error message too like `"Energy must be between ${minimumEnergy} and ${maximumEnergy}"` for example

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/cardsapi/resources/CardsResource.groovy:
```diff
@@ -65,4 +82,129 @@ class CardsResource extends Resource {
+    @Path('')
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    Response postCard (@Valid Card newCard) {
```
In your swagger spec, it looks like you are posting a NewData object (like the resultobject) but that's not the case here. We should be accepting a resultobject as the Post body to adhere to jsonapi spec.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/cardsapi/db/CardDAO.groovy:
```diff
@@ -71,8 +35,29 @@ interface CardDAO extends Closeable {
+            (:description)
+        )
+        """)
+    void postCard(@Bind("id") Integer id,
```
Instead of binding all the individual objects, you could have one parameter as the Card object and use `@BindBean` to bind an object to the SQL query's bind variables

ianbrown9475 committed:
Change POST to use resultObject

ianbrown9475 commented:
- POST /cards endpoint now takes in a resultObject instead of a Card
- energyMin and energyMax now class variables
- BindBean annotation used for simplicity

ianbrown9475 committed:
Merge branch 'master' into feature/CO-1116-POST/cards_endpoint

ianbrown9475 committed:
Remove unused dependencies

jaredkosanovic commented:
I tried POSTing this:
```
{
  "data": {
    "attributes": {
      "type": "skill",
      "name": "Defend",
      "color": "red",
      "rarity": "basic",
      "energy": 1,
      "description": "Gain 5 Block."
    }
  }
}
```
and got this response:
```
{
  "status": 500,
  "developerMessage": "Uncaught exception: groovy.lang.MissingPropertyException",
  "userMessage": "The application ran into an internal error. If the problem persists, please contact application support.",
  "code": 1500,
  "details": "https://developer.oregonstate.edu/documentation/error-reference#1500"
}
```
snippet from stack trace:
```
ERROR [2018-06-11 15:38:52,998] edu.oregonstate.mist.api.jsonapi.GenericExceptionMapper: Uncaught exception
! groovy.lang.MissingPropertyException: No such property: regEx for class: edu.oregonstate.mist.cardsapi.resources.CardsResource
! at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:51)
! at org.codehaus.groovy.runtime.callsite.PogoGetPropertySite.getProperty(PogoGetPropertySite.java:49)
! at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGroovyObjectGetProperty(AbstractCallSite.java:304)
! at edu.oregonstate.mist.cardsapi.resources.CardsResource.resultObjectValidator(CardsResource.groovy:186)
! at edu.oregonstate.mist.cardsapi.resources.CardsResource$resultObjectValidator$1.callCurrent(Unknown Source)
! at edu.oregonstate.mist.cardsapi.resources.CardsResource.postCard(CardsResource.groovy:157)
```

jaredkosanovic commented:
Also, posting nothing in the request body will result in a null pointer exception:
```
{
  "status": 500,
  "developerMessage": "Uncaught exception: java.lang.NullPointerException",
  "userMessage": "The application ran into an internal error. If the problem persists, please contact application support.",
  "code": 1500,
  "details": "https://developer.oregonstate.edu/documentation/error-reference#1500"
}
```

Could you also put an example POST body in the readme as an example for creating a new card?

ianbrown9475 committed:
Fix POST, add example POST body to readme

ianbrown9475 commented:
- CardsResource uses validPattern variable instead of nonexistent regEx variable
- Added example POST body to readme
- Added error checking for empty POST body

ianbrown9475 committed:
Check types in POST body

ianbrown9475 commented:
- Types of fields in POST body are now checked
- Null fields in POST body are now checked
